File: apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/SecretRedactorTest.kt
package com.agentic.wallet.agent.provider

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class SecretRedactorTest {
    @Test
    fun replacesExactSecret() {
        val secret = "sk-test-1234567890ABCDEF"
        val redacted = SecretRedactor.redact("Authorization failed for $secret", secret)
        assertFalse(redacted.contains(secret))
        assertTrue(redacted.contains("[redacted]"))
    }

    @Test
    fun trimsSecretBeforeReplacing() {
        val redacted = SecretRedactor.redact("Failed: foo-secret", "  foo-secret  ")
        assertEquals("Failed: [redacted]", redacted)
    }

    @Test
    fun noOpOnBlankSecret() {
        assertEquals("hello world", SecretRedactor.redact("hello world", ""))
        assertEquals("hello world", SecretRedactor.redact("hello world", null))
        assertEquals("hello world", SecretRedactor.redact("hello world", "   "))
    }

    @Test
    fun redactsBearerPattern() {
        val redacted = SecretRedactor.redact("got Bearer abcDEF.123-456_xyz token", null)
        assertTrue(redacted.contains("Bearer [redacted]"))
        assertFalse(redacted.contains("abcDEF.123-456_xyz"))
    }

    @Test
    fun redactsSkProjPattern() {
        val redacted = SecretRedactor.redact("error sk-proj-ABCDEF12345 occurred", null)
        assertTrue(redacted.contains("sk-proj-[redacted]"))
        assertFalse(redacted.contains("sk-proj-ABCDEF12345"))
    }

    @Test
    fun redactsKeyValueStyle() {
        val redacted = SecretRedactor.redact("api-key=ABCDEFGHIJKLMNOP", null)
        assertTrue("Expected key redaction, got $redacted", redacted.contains("[redacted]"))
        assertFalse(redacted.contains("ABCDEFGHIJKLMNOP"))
    }

    @Test
    fun redactsJwtLikePattern() {
        val jwt = "eyJabcd1234567890abcd.eyJsubbing0987654321abc.signaturedeadbeefcafe1234"
        val redacted = SecretRedactor.redact("got token $jwt then proceeded", null)
        assertTrue("Expected JWT redaction, got: $redacted", redacted.contains("[redacted-token]"))
        assertFalse(redacted.contains(jwt))
    }

    @Test
    fun redactsBareSkPattern() {
        val redacted = SecretRedactor.redact("error with sk-ABCDEF1234567890XYZ occurred", null)
        assertTrue(redacted.contains("sk-[redacted]"))
        assertFalse(redacted.contains("ABCDEF1234567890XYZ"))
    }

    @Test
    fun redactsMultipleSecretOccurrences() {
        val secret = "sk-test-multi-EXAMPLEKEY"
        val redacted = SecretRedactor.redact(
            "first $secret then again $secret done",
            secret,
        )
        assertFalse("Both occurrences should be redacted: $redacted", redacted.contains(secret))
        // both occurrences replaced — at least two [redacted] markers
        assertEquals(2, "\\[redacted\\]".toRegex().findAll(redacted).count())
    }

    @Test
    fun regexFallbacksApplyEvenWithoutExactSecret() {
        val redacted = SecretRedactor.redact("calling Bearer abc123XYZ_456.token-data more text", null)
        assertTrue(redacted.contains("Bearer [redacted]"))
        assertFalse(redacted.contains("abc123XYZ_456.token-data"))
    }
}

File: apps/browser-demo/src/deviceAgent/__tests__/secretRedactor.test.ts
// Ported from
// apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/SecretRedactorTest.kt.
// Pins each redactor regex against the same inputs the Kotlin tests use so
// drift between android-native and browser-native error surfaces is impossible.

import { describe, expect, it } from 'vitest';

import { redactSecret } from '../provider/secretRedactor.js';

describe('redactSecret', () => {
  it('replaces the exact secret when provided', () => {
    const secret = 'sk-test-1234567890ABCDEF';
    const redacted = redactSecret(`Authorization failed for ${secret}`, secret);
    expect(redacted.includes(secret)).toBe(false);
    expect(redacted.includes('[redacted]')).toBe(true);
  });

  it('trims the secret before replacing', () => {
    expect(redactSecret('Failed: foo-secret', '  foo-secret  ')).toBe('Failed: [redacted]');
  });

  it('is a no-op when the secret is blank, null, or undefined', () => {
    expect(redactSecret('hello world', '')).toBe('hello world');
    expect(redactSecret('hello world', null)).toBe('hello world');
    expect(redactSecret('hello world', undefined)).toBe('hello world');
    expect(redactSecret('hello world', '   ')).toBe('hello world');
  });

  it('redacts the Bearer pattern', () => {
    const redacted = redactSecret('got Bearer abcDEF.123-456_xyz token', null);
    expect(redacted.includes('Bearer [redacted]')).toBe(true);
    expect(redacted.includes('abcDEF.123-456_xyz')).toBe(false);
  });

  it('redacts the sk-proj- pattern', () => {
    const redacted = redactSecret('error sk-proj-ABCDEF12345 occurred', null);
    expect(redacted.includes('sk-proj-[redacted]')).toBe(true);
    expect(redacted.includes('sk-proj-ABCDEF12345')).toBe(false);
  });

  it('redacts the bare sk- pattern', () => {
    const redacted = redactSecret('error with sk-ABCDEF1234567890XYZ occurred', null);
    expect(redacted.includes('sk-[redacted]')).toBe(true);
    expect(redacted.includes('ABCDEF1234567890XYZ')).toBe(false);
  });

  it('redacts a JWT-shaped token', () => {
    const jwt = 'eyJabcd1234567890abcd.eyJsubbing0987654321abc.signaturedeadbeefcafe1234';
    const redacted = redactSecret(`got token ${jwt} then proceeded`, null);
    expect(redacted.includes('[redacted-token]')).toBe(true);
    expect(redacted.includes(jwt)).toBe(false);
  });

  it('redacts api-key/token/secret key=value style', () => {
    const redacted = redactSecret('api-key=ABCDEFGHIJKLMNOP', null);
    expect(redacted.includes('[redacted]')).toBe(true);
    expect(redacted.includes('ABCDEFGHIJKLMNOP')).toBe(false);
  });

  it('redacts every occurrence when the secret appears multiple times', () => {
    const secret = 'sk-test-multi-EXAMPLEKEY';
    const redacted = redactSecret(`first ${secret} then again ${secret} done`, secret);
    expect(redacted.includes(secret)).toBe(false);
    const matches = redacted.match(/\[redacted\]/g);
    expect(matches?.length ?? 0).toBe(2);
  });

  it('applies the regex fallback when no exact secret is supplied', () => {
    const redacted = redactSecret('calling Bearer abc123XYZ_456.token-data more text', null);
    expect(redacted.includes('Bearer [redacted]')).toBe(true);
    expect(redacted.includes('abc123XYZ_456.token-data')).toBe(false);
  });

  it('uses the with-space KEY_VALUE form (api key=…, not just api-key=…)', () => {
    // Kotlin SecretRedactor.kt uses [-_ ]? (with space). A copy that loses the
    // space will let "api key=…" through. Pin both spellings.
    const redacted = redactSecret('api key=ABCDEFGHIJKLMNOP', null);
    expect(redacted.includes('[redacted]')).toBe(true);
    expect(redacted.includes('ABCDEFGHIJKLMNOP')).toBe(false);
  });
});
